### PR TITLE
Transfer code

### PR DESCRIPTION
**Pull request by KaiNissen**

### CODE REVIEW

File: src/Store/Factory.php [hunk 1]
```diff
@@ -38,6 +41,14 @@ public function newInstaller() {
...
 	 */
 	public function getEntityManager() {
+		if ( !$this->entityManager ) {
+			$this->setupEntityManager();
+		}
+
+		return $this->entityManager;
+	}
+
+	public function setupEntityManager() {
```
Comment (JeroenDeDauw) [inline src/Store/Factory.php]: Does this need to be public?

File: src/Store/Factory.php [hunk 2]
```diff
@@ -51,12 +62,14 @@ public function getEntityManager() {
...
 
+		$eventManager->addEventListener( 'prePersist', new TransferCodeListener( new TransferCodeGenerator() ) );
+
 		$entityManager = EntityManager::create( $this->connection, $config, $eventManager );
 
 		$platform = $entityManager->getConnection()->getDatabasePlatform();
 		$platform->registerDoctrineTypeMapping( 'enum', 'string' );
 
-		return $entityManager;
+		$this->entityManager = $entityManager;
```
Comment (JeroenDeDauw) [inline src/Store/Factory.php]: I'd return the created EM and have the caller set it to a field, as this is also where the field is used

File: src/Store/TransferCodeListener.php [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	public function prePersist( LifecycleEventArgs $args ) {
+		$entity = $args->getObject();
+		$entityManager = $args->getObjectManager();
+
+		if ( $entity instanceof Donation && !empty( $entity->getTransferCode() ) ) {
+			$donationRepository = $entityManager->getRepository( Donation::class );
+			while ( true ) {
+				$dataSets = $donationRepository->findBy( [ 'transferCode' => $entity->getTransferCode() ] );
+
+				if ( count( $dataSets ) === 0 ) {
```
Comment (JeroenDeDauw) [inline src/Store/TransferCodeListener.php]: if ( empty( $dataSets ) ) {

File: src/Store/TransferCodeListener.php [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+					break;
+				}
```
Comment (JeroenDeDauw) [inline src/Store/TransferCodeListener.php]: "loop until" more appropriate no?

**Comment by JeroenDeDauw** [comment]
Looks reasonable enough to me, though then I really don't know what this is all about, so can't really review it.

**Comment by JeroenDeDauw** [comment]
I've looked at this again now and would like to talk about this in person.

**Comment by KaiNissen** [comment]
factory fix in #54; transfer code generation will reside in wmde/FundraisingFrontend, so closing this one

**Pull request closed without merging**